File: editor/ARCed/src/Core/ARC_Dump.py
'''
Created on Sep 4, 2011

ARC_Dump: dumps data into msgpack structured files
'''
import types

import msgpack

__io = None
__symbols = {}
__strings = {}
__ids = {}
__stack = []
__finaldump = {}
__sym_count = 0
__pending_objects = []
__post_load_objects = []
__class_path_redirects = {}

def dump(obj, io, redirects={}):
    global __io
    global __class_path_redirects
    __class_path_redirects = redirects
    __io = io
    queue_dump_object(obj)
    start_dump()
    write_dump()
    reset()
  
  
def load(io, redirects={}):
    global __io
    global __symbols
    global __class_path_redirects
    __class_path_redirects = redirects
    __io = io
    build_object_table()
    link_objects()
    call_arc_post_load()
    object = __symbols[0]
    if object[0] >= 0 and object[0] <= 4: #False, True, Nil, String, Numeric
        return_obj = object[1]
    elif object[0] >= 5 and object[0] <= 7:#Array, Hash, Class
        return_obj = object[1][0]
    else:
        return_obj = None
    reset()
    return return_obj
  
  
def build_object_table():
    global __io
    global __symbols
    global __pending_objects
    global __class_path_redirects
    load_string = __io.read()
    objects = msgpack.unpackb(load_string)
    for key, item in objects:
        __symbols[key] = item 
        if item[0] >= 0 and item[0] <= 4: #False, True, Nil, String, Numeric
            continue
        elif item[0] >= 5 and item[0] <= 6: #Array, Hash
            if len(item[1]) > 0:
                __pending_objects.append(key) 
        elif item[0] == 7:    #Class
            if __class_path_redirects.has_key(item[1][0]):
                klass_path = __class_path_redirects[item[1][0]]
            else:
                klass_path =item[1][0]
            klasses = klass_path.split("::")
            if globals().has_key(klasses[0]):
                klass = globals()[klasses[0]]
            else:
                raise TypeError, "No class defined: %s} " % klasses[0]
            if len(klasses) > 1:
                for index in range(len(klasses)):
                    if index == 0:
                        continue
                    if klass.__dict__.has_key(klasses[index]):
                        klass = klass.__dict__[klasses[index]]
                    else:
                        raise TypeError, "No class defined: %s} " % klasses[index]
            klass_obj = klass.__new__(klass)
            item[1][0] = klass_obj
            __pending_objects.append(key)
        else:
            raise TypeError, "Unrecognized type declaration: %s" % type
  
  
def link_objects():
    global __symbols
    global __pending_objects
    global __post_load_objects
    for obj_key in __pending_objects:
        obj = __symbols[obj_key]
        if obj[0] == 5: #array
            new_array = obj[1][0]
            for item in obj[1][1]:
                if item[0] == 0: #actualy object
                    new_array.append(item[1])
                elif item[0] == 1: #link to array, hash, or user object
                    new_array.append(__symbols[item[1]][1][0])
                elif item[0] == 2: #string
                    new_array.append(__symbols[item[1]][1])
        elif obj[0] == 6: #hash
            new_hash = obj[1][0]
            for pair in obj[1][1]:
                if pair[0][0] == 0: #actualy object
                    key = pair[0][1]
                elif pair[0][0] == 1: #link to array, hash, or user object
                    key = __symbols[pair[0][1]][1][0]
                elif pair[0][0] == 2: #string
                    key = __symbols[pair[0][1]][1]
                if pair[1][0] == 0: #actualy object
                    value = pair[1][1]
                elif pair[1][0] == 1: #link to array, hash, or user object
                    value = __symbols[pair[1][1]][1][0]
                elif pair[1][0] == 2: #string
                    value = __symbols[pair[1][0]][1]
                new_hash[key] = value    
        elif obj[0] == 7: #nonstandard object
            klass_obj = obj[1][0]
            if hasattr(klass_obj, "_arc_load"):
                klass_obj._arc_load(obj[1][1])
            else:
                for pair in obj[1][1]:
                    variable = pair[0]
                    if pair[1][0] == 0: #actualy object
                        value = pair[1][1]
                    elif pair[1][0] == 1: #link to array, hash, or user object
                        value = __symbols[pair[1][1]][1][0]
                    elif pair[1][0] == 2: #string
                        value = __symbols[pair[1][1]][1]
                    setattr(klass_obj, variable, value)
            if hasattr(klass_obj, "_post_arc_load"):
                __post_load_objects.append(klass_obj)
        
      
  
  
def call_arc_post_load():
    global __post_load_objects
    for obj in __post_load_objects:
        obj._post_arc_load()
  
def reset():
    global __io
    global __symbols
    global __strings
    global __ids
    global __stack
    global __finaldump
    global __sym_count
    global __pending_objects
    global __post_load_objects
    global __class_path_redirects
    __io = None
    __symbols = {}
    __strings = {}
    __ids = {}
    __stack = []
    __finaldump = {}
    __sym_count = 0
    __pending_objects = []
    __post_load_objects = []
    __class_path_redirects = {}
  
  
def start_dump():
    global __stack
    while len(__stack) > 0:
        dump_object(__stack.pop(0))
    
  
  
def write_dump():
    global __io
    global __finaldump
    dump_string = msgpack.packb(__finaldump)
    __io.write(dump_string)
  
  
def queue_dump_object(obj):
    if obj is False:
        result = dump_false_object(obj)
    elif obj is True:
        result = dump_true_object(obj)
    elif obj is None:
        result = dump_nil_object(obj)
    elif isinstance(obj, (types.IntType, types.LongType, types.FloatType)):
        result = dump_numeric_object(obj)
    elif isinstance(obj, types.StringType):
        result = dump_string_object(obj)
    elif isinstance(obj, (types.ListType, types.TupleType)):
        result = queue_dump_array_object(obj)
    elif isinstance(obj, types.DictType):
        result = queue_dump_hash_object(obj)
    else:
        result = queue_dump_nonstandard_object(obj)
    return result
  
  
def dump_object(obj):
    if isinstance(obj, (types.ListType, types.TupleType)):
        dump_array_object(obj)
    elif isinstance(obj, types.DictType):
        dump_hash_object(obj)
    else:
        dump_nonstandard_object(obj)
    
  
  
def dump_true_object(obj):
    return [0, True]
  
  
def dump_false_object(obj):
    return [0, False]
  
  
def dump_nil_object(obj):
    return [0, None]
  
  
def dump_numeric_object(obj):
    return [0, obj]
  
  
def dump_string_object(obj):
    global __strings
    global __ids
    if not __strings.has_key(obj):
        __strings[obj] = id(obj)
        dump_array = enter_obj(obj)
        dump_array[0], dump_array[1] = 4, obj
    return [2, __ids[__strings[obj]]]
  
  
def queue_dump_array_object(obj):
    global __symbols
    global __ids
    if not __symbols.has_key(id(obj)):
        __stack.append(obj)
        dump_array = enter_obj(obj)
        dump_array[0], dump_array[1] = 5, [[],[]]
    return [1, __ids[obj.object_id]]
  
  
def dump_array_object(obj):
    global __symbols
    dump_array = __symbols[id(obj)][1]
    for value in obj:
        dump_array[1][1].append(queue_dump_object(value))
  
  
def queue_dump_hash_object(obj):
    global __symbols
    global __ids
    if not __symbols.has_key(id(obj)):
        __stack.append(obj)
        dump_array = enter_obj(obj)
        dump_array[0], dump_array[1] = 6, [{}, []]
    return [1, __ids[id(obj)]]
  
 
def dump_hash_object(obj):
    global __symbols
    dump_array = __symbols[obj.object_id][1]
    for key, value in obj:
        pair = [queue_dump_object(key), queue_dump_object(value)]
        dump_array[1][1].append(pair)
        
        
def request_class_path(obj):
    global __class_path_redirects
    if hasattr(obj, "_arc_class_path"):
        klass_path = obj._arc_class_path
    else:
        klass_path = "%s::%s" % (obj.__class__.__module__, obj.__class__.__name__)
    if __class_path_redirects.has_key(klass_path):
        klass_path = __class_path_redirects[klass_path]
    return klass_path
    
    
def queue_dump_nonstandard_object(obj):
    global __symbols
    global __ids
    if not __symbols.has_key(id(obj)):
        __stack.append(obj)
        dump_array = enter_obj(obj)
        klass = request_class_path(obj)
        dump_array[0], dump_array[1] = 7, [klass,[]]
    return [1, __ids[id(obj)]]

def request_instance_variables(obj):
    if hasattr(obj, "_arc_instance_variables"):
        return obj._arc_instance_variables
    else:
        list = []
        for key, value in obj.__dict__:
            if key[0] is not "_":
                if not isinstance(value, (types.FunctionType, types.ClassType, types.MethodType, 
                                          types.ModuleType, types.SliceType, types.LambdaType, 
                                          types.GeneratorType)):
                    list.append(key)
        return list
  
def dump_nonstandard_object(obj):
    global __symbols
    dump_array = __symbols[obj.object_id][1]
    if hasattr(obj, "_prep_arc_dump"):
        obj._prep_arc_dump()
    if hasattr(obj, "_arc_dump"):
        dump_array[1][1] = obj._arc_dump()
    else:
        if hasattr(obj, "_arc_exclude"):
            excludes = obj._arc_exclude
        else:
            excludes = []
        instance_variables = request_instance_variables(obj)
        for value in instance_variables:
            if value in excludes:
                continue
            pair = [value, queue_dump_object(getattr(obj, value))]
            dump_array[1][1].append(pair)
    
  
 
def enter_obj(obj):
    global __symbols
    global __ids
    global __finaldump
    global __sym_count
    key = obj.object_id
    dump_array = [None, None]
    __symbols[key] = [obj, dump_array]
    __ids[key] = __sym_count
    __finaldump[__sym_count] = dump_array
    __sym_count += 1
    return dump_array
  

File: editor/ARCed/tests/ARC_dump/arc_dump.rb
load 'data.rb'
require 'msgpack'

module ARC_Dump

  @@io = nil
  @@symbols = {}
  @@strings = {}
  @@ids = {}
  @@stack = []
  @@finaldump = {}
  @@sym_count = 0
  @@pending_objects = []
  @@post_load_objects = []
  @@class_path_redirects = {}

  def self.dump(obj, io, redirects = {})
    @@class_path_redirects = redirects
    @@io = io
    self.queue_dump_object(obj)
    self.start_dump
    self.write_dump
    self.reset
  end
  
  def self.load(io, redirects = {})
    @@class_path_redirects = redirects
    @@io = io
    self.build_object_table
    self.link_objects
    self.call_arc_post_load
    object = @@symbols[0]
    case object[0]
    when 0..4 #False, True, Nil, String, Numeric
      return_obj = object[1]
    when 5..7 #Array, Hash, Class
      return_obj = object[1][0]
    else
      return_obj = nil
    end
    self.reset
    return return_obj
  end
  
  def self.build_object_table
    load_string = @@io.read()
    objects = MessagePack.unpack(load_string)
    objects.each {|key, item|
      @@symbols[key] = item 
      case item[0]
      when 0..4 #False, True, Nil, String, Numeric
        next
      when 5..6 #Array, Hash
        @@pending_objects.push(key) if !item[1].empty?
      when 7    #Class
        if @@class_path_redirects.has_key?(item[1][0])
          klass_path = @@class_path_redirects[item[1][0]]
        else
          klass_path = item[1][0]
        end
        klasses = klass_path.split("::")
        if Kernel.const_defined?(klasses[0].to_sym)
          klass = Kernel.const_get(klasses[0].to_sym)
        else
          raise TypeError, "No class defined: #{klasses[0]} "
        end
        if klasses.size > 1
          klasses.each_index {|index|
            next if index == 0
            klass = klass.const_get(klasses[index].to_sym)
          }
        end
        klass_obj = klass.allocate()
        item[1][0] = klass_obj
        @@pending_objects.push(key)
      else
        raise TypeError, "Unrecognized object declaration: #{item}"
      end
    } 
  end
  
  def self.link_objects
    @@pending_objects.each {|obj_key|
      obj = @@symbols[obj_key]
      case obj[0]
      when 5
        new_array = obj[1][0]
        obj[1][1].each {|item|
          case item[0]
          when 0 #actualy object
            new_array.push(item[1])
          when 1 #link to array, hash, or user object
            new_array.push(@@symbols[item[1]][1][0])
          when 2 #string
            new_array.push(@@symbols[item[1]][1])
          end
        }
      when 6
        new_hash = obj[1][0]
        obj[1][1].each {|pair|
          case pair[0][0]
          when 0 #actualy object
            key = pair[0][1]
          when 1 #link to array, hash, or user object
            key = @@symbols[pair[0][1]][1][0]
          when 2 #string
            key = @@symbols[pair[0][1]][1]
          end
          case pair[1][0]
          when 0 #actualy object
            value = pair[1][1]
          when 1 #link to array, hash, or user object
            value = @@symbols[pair[1][1]][1][0]
          when 2 #string
            value = @@symbols[pair[1][1]][1]
          end
          new_hash[key] = value
        }
      when 7
        klass_obj = obj[1][0]
        if klass_obj.class.method_defined?("_arc_load")
          klass_obj._arc_load(obj[1][1])
        else
          obj[1][1].each {|pair|
            variable = pair[0]
            case pair[1][0]
            when 0 #actualy object
              value = pair[1][1]
            when 1 #link to array, hash, or user object
              value = @@symbols[pair[1][1]][1][0]
            when 2 #string
              value = @@symbols[pair[1][1]][1]
            end
            klass_obj.instance_variable_set(("@" + variable.strip).to_sym, value)
          }
        end
        if klass_obj.class.method_defined?("_post_arc_load")
          @@post_load_objects.push(klass_obj)
        end
      end
    }
  end
  
  def self.call_arc_post_load
    @@post_load_objects.each {|obj|
      obj._post_arc_load
    }
  end
  
  def self.reset
    @@io = nil
    @@symbols = {}
    @@strings = {}
    @@ids = {}
    @@stack = []
    @@finaldump = {}
    @@sym_count = 0
    @@pending_objects = []
    @@class_path_redirects = {}
  end
  
  def self.start_dump
    while !@@stack.empty?
      self.dump_object(@@stack.shift)
    end
  end
  
  def self.write_dump
    dump_string = @@finaldump.to_msgpack
    @@io.write(dump_string)
  end
  
  def self.queue_dump_object(obj)
    if obj.is_a?(FalseClass)
      result = self.dump_false_object(obj)
    elsif obj.is_a?(TrueClass)
      result = self.dump_true_object(obj)
    elsif obj.is_a?(NilClass)
      result = self.dump_nil_object(obj)
    elsif obj.is_a?(Numeric)
      result = self.dump_numeric_object(obj)
    elsif obj.is_a?(String)
      result = self.dump_string_object(obj)
    elsif obj.is_a?(Array)
      result = self.queue_dump_array_object(obj)
    elsif obj.is_a?(Hash)
      result = self.queue_dump_hash_object(obj)
    else
      result = self.queue_dump_nonstandard_object(obj)
    end
    return result
  end
  
  def self.dump_object(obj)
    if obj.is_a?(Array)
      self.dump_array_object(obj)
    elsif obj.is_a?(Hash)
      self.dump_hash_object(obj)
    else
      self.dump_nonstandard_object(obj)
    end
  end
  
  def self.dump_true_object(obj)
    return [0, true]
  end
  
  def self.dump_false_object(obj)
    return [0, false]
  end
  
  def self.dump_nil_object(obj)
    return [0, nil]
  end
  
  def self.dump_numeric_object(obj)
    return [0, obj]
  end
  
  def self.dump_string_object(obj)
    if !@@strings.has_key?(obj)
      @@strings[obj] = obj.object_id
      dump_array = self.enter_obj(obj)
      dump_array[0], dump_array[1] = 4, obj
    end
    return [2, @@ids[@@strings[obj]]]
  end
  
  def self.queue_dump_array_object(obj)
    if !@@symbols.has_key?(obj.object_id)
      @@stack.push(obj)
      dump_array = self.enter_obj(obj)
      dump_array[0], dump_array[1] = 5, [[],[]]
    end
    return [1, @@ids[obj.object_id]]
  end
  
  def self.dump_array_object(obj)
    dump_array = @@symbols[obj.object_id][1]
    obj.each {|value| 
      dump_array[1][1].push(self.queue_dump_object(value))
    }
  end
  
  def self.queue_dump_hash_object(obj)
    if !@@symbols.has_key?(obj.object_id)
      @@stack.push(obj)
      dump_array = self.enter_obj(obj)
      dump_array[0], dump_array[1] = 6, [{}, []]
    end
    return [1, @@ids[obj.object_id]]
  end
 
  def self.dump_hash_object(obj)
    dump_array = @@symbols[obj.object_id][1]
    obj.each_pair {|key, value|
      pair = [self.queue_dump_object(key), self.queue_dump_object(value)]
      dump_array[1][1].push(pair)
    }
  end
 
  def self.queue_dump_nonstandard_object(obj)
    if !@@symbols.has_key?(obj.object_id)
      @@stack.push(obj)
      dump_array = self.enter_obj(obj)
      if @@class_path_redirects.has_key?(obj.class.to_s)
        klass_path = @@class_path_redirects[obj.class.to_s]
      else
        klass_path = obj.class.to_s
      end
      dump_array[0], dump_array[1] = 7, [klass_path,[]]
    end
    return [1, @@ids[obj.object_id]]
  end
  
  def self.dump_nonstandard_object(obj)
    dump_array = @@symbols[obj.object_id][1]
    if obj.class.method_defined?("_prep_arc_dump")
      obj._prep_arc_dump
    end
    if obj.class.method_defined?("_arc_dump")
      dump_array[1][1] = obj._arc_dump
    else
      if obj.class.method_defined?("_arc_exclude")
        excludes = obj._arc_exclude
      else
        excludes = []
      end
      instance_variables = obj.instance_variables
      instance_variables.each {|value|
        name = value.to_s.gsub("@", "")
        next if excludes.include?(name)
        pair = [name, self.queue_dump_object(obj.instance_variable_get(value))]
        dump_array[1][1].push(pair)
      }
    end
  end
 
  def self.enter_obj(obj)
    key = obj.object_id
    dump_array = [nil,nil]
    @@symbols[key] = [obj, dump_array]
    @@ids[key] = @@sym_count
    @@finaldump[@@sym_count] = dump_array
    @@sym_count += 1
    return dump_array
  end

end
